templates/repo/diff/image_diff.tmpl

```diff
@@ -74,7 +74,9 @@
 				</div>
 				{{end}}
 			</div>
-			<div class="ui active centered inline loader gt-mb-4"></div>
+			<div class="img-diff-spinner gt-df gt-jc gt-py-4">
+				{{template "shared/spinner" dict "Size" 64}}
+			</div>
```

Comment:
The "spinner" doesn't seem to help too much.


IMO:

```
<div class="img-diff-spinner is-loading gt-py-4" style="height: 40px"></div>
```

is clearer and simpler than these 3 lines.

Reply:
Why duplicate 3 times? I think shared template is much superior. Div and class can be templated too.

Reply:
I think shared template will be a good option here.

Reply:
Yes, ideally all such things should be subtemplated, it makes them easily modifyable in one place.

Reply:
> Yes, ideally all such things should be subtemplated, it makes them easily maintainable in one place.

I have objection for doing so and I disagree about making everything (checkbox/button/etc) into a subtemplate.

Think about a real case:

```

{{template "shared/button" dict "class" "my-button" "title" "my-title"}}
{{template "shared/spinner" dict "class" "my-other-class" "size" "40"}}
```

vs

```
<button class="my-class">my-title</button>
<div class="is-loading my-other-class" style="height: 40px"></div>
```

I would say the HTML ones are clearer than the over-abstracted subtemplates. For the "subtemplates", even worse, you will have to add more and more attributes like `disabled`, `readony`, `type`, etc. In the end, you added every HTML attribute into the subtemplate.

And the raw HTML code benefits from:

1. Clear code, every developer knows HTML
2. Code highlight
3. IDE (I am using GoLand) code-completion
4. HTML template lint

Reply:
If you think there is no difference, I also consider there is no difference of using HTML directly, instead of introducing a new one-line subtemplate, what's the difference?

Reply:
The difference is future exensibilty and doing changes in 1 place instead of many. Take for example if we want to replace the inline style in the future with CSS `attr` (currently unsupported by browsers), with the subtemplate, it's a 1-line change in templates instead of currently 3-line.

Reply:
> The difference is future exensibilty and doing changes in 1 place instead of many,

I do not see how it improves the HTML code, in the end, for simple one line HTML code block, it just re-invents the syntax:

>>  For the "subtemplates", even worse, you will have to add more and more attributes like `disabled`, `readonly`, `type`, etc. In the end, you will add every HTML attribute into the subtemplate and inventing a "new template-HTML" syntax.

And for the possible `attr`:

> Take for example if we want to replace the inline style in the future with CSS attr (currently unsupported by browsers): With the subtemplate, it's a 1-line change in templates instead of currently 3-line.

At the moment I do not see the necessity to use inline style. The height could be set by some general height like `gt-height-4em` / `gt-height-12em`, or just use the parent container's height. ps: Although tailwind have similar classes https://tailwindcss.com/docs/height, but I do not think the tailwind styles are clear for this case (it doesn't seem intuitive to use `h-4` for `1rem` / `16px`)

But the "subtemplate" causes other problems, for example: why always "gt-dib"? How to add more classes? And it always needs an extra parent container, which could be avoided in most cases by inline HTML.

----

To answer the "difference" above:

>> Clear code, every developer knows HTML
> Same with subtemplates, just some indirection to DRY the code

I guess most developers are not familiar with the "template" syntax. One-line HTML doesn't mean "repeat", otherwise copying the one-line template code could also be considered as "repeat".

>> Code highlight
> No difference, subtemplate is highlighted as well

By using raw HTML, it could clearly see the attributes, values and the content. But by using subtemplates, everything is a simple map value string.

>> IDE (I am using GoLand) code-completion
>If golang would provide built-in `dict`-like functionality in templates, editors could auto-complete the available variables from the subtemplate. I think `dict` is by far the best way to have extensible subtemplates, it works and feels the same like for example react components that take a options object.

Yup, `dict` is the only approach at the moment, but it doesn't read like React or HTML.

>> HTML template lint
> Not difference, subtemplates is linted as well

Passing the values by `dict` make it difficult to lint, and difficult to grep or search CSS classes.

Reply:
> For the "subtemplates", even worse, you will have to add more and more attributes like disabled, readonly, type,

In React one would just `<div {...props}>` to spread all props as-is onto the element, but I think go template can not do such things sadly because you have no way of knowing which props to not spread without a formal dict-argument definition on the template, e.g. like function args.